\documentclass[12pt,addpoints]{exam}
% uncomment for exam with solutions
\printanswers
%\pointsinmargin
%\marginpointname{ pts}
\bracketedpoints
%\pagestyle{empty}
%\qformat{Question \thequestion\dotfill \emph{\totalpoints\ points}}
\usepackage{graphicx}
\usepackage{comment}
\usepackage{amsthm}
\usepackage{color}
\usepackage{verbatim}
\usepackage{amsmath}
\usepackage{amssymb}

\newcommand{\bX}{\ensuremath{\boldsymbol{X}}}

\begin{document}
\SweaveOpts{concordance=TRUE}

\begin{center}
\section*{Homework 1}
\end{center}

\begin{questions}

\question Out of the pressure of the classroom, let's explore the modeling exercise we did in class. \\
\textbf{Knowledge:} Assume all transcripts human cells transcribe are known. \\
\textbf{Hypothesis and Question:}
A human cell of type A has different transcript levels from a human cell of type B (\textit{e.g.}~one cancerous, other not), (and these transcript levels can reveal the cause of cancer).  Which transcripts are differentially expressed? \\
\textbf{Experiment:}
\begin{itemize}
\item Materials: one cell of type A, one cell of type B.
\item Randomly sample $n$ mRNA from each cell.
\item Unambiguously identify the transcript of each mRNA.
\end{itemize}
The data are counts $\bX_i = (X_{i1},X_{i2},\ldots,X_{im})$, where $i\in\{A,B\}$ indexes the cell and $m$ is the total number of possible transcripts.
We will focus on the single gene $j$, though one could also model the entire multivariate vector.
Notice, that the components of the vectors are not independent: as one count, say $X_{ij}$ increases, other counts $X_{ik}, k\neq j$ will tend to decrease because the total number of transcripts in the cell is finite.
Fortunately, if $m$ is large and more specifically many members of vector $\bX_i$ are non-zero, then the components of the vector are nearly independent.
Bioinformaticians make use of this fact all the time in RNA-seq experiments; it allows them to handle one gene at a time.
\begin{parts}
\part If cell $A$ has $T_A$ total mRNAs and cell $B$ has $T_B$ total mRNAs, not necessarily equal, and the sample size $n$ is \textit{not} much smaller than $T_A$ (or $T_B$) and sampling is without replacement, specify a distribution for the data under $H_0$.  Continue to assume this model until directed not to.
\\ \textbf{Solution:}
\[ X_{ij} \sim H(T_i, \theta_{ij}, n), \]
where $H(\cdot)$ is the hypergeometric distribution and $\theta_{ij}$ is the unknown parameter representing the number of transcript $j$ in cell $i \in \{A,B\}$.
\part Hypothesis testing consists of several parts (you can read more about it in the Stat 342 notes):
	\begin{itemize}
	\item A \textbf{model}, which you have already specified.
	\item The \textbf{null} ($H_0$) and \textbf{alternative} ($H_a$) hypotheses.
	\item A {\bf test statistic}, $T(\bX)$, some function of the data $\bX$, and a sampling distribution for the $T(\bX)$ under $H_0$.
	\item A {\bf significance level} ($\alpha$).
	\item A {\bf rejection region} ($R_R$).
	\end{itemize}
The null hypothesis $H_0$ is a statement about population parameters in your model.
When there are no differentially expressed genes in the two cells, what is true about the parameters of the distribution you named, \textit{i.e.}~what is $H_0$?
\\ \textbf{Solution:}
\[H_0: \frac{\theta_{Aj}}{T_A} = \frac{\theta_{Bj}}{T_B}\]
\part Since $H_0$ is a statement about population parameters, a frequently useful test statistic is the maximum likelihood estimator (MLE) of the involved parameters, especially since the MLE is functionally invariant meaning that the MLE of some function of a parameter, $\widehat{f(\theta)}$, is in most cases (see Stat 342 for details), the function applied to the MLE of the parameter: $f(\hat\theta)$, where $\hat\cdot$ is used to indicate the MLE.
\part Is it possible to obtain MLEs of the parameters in your model?  (Don't attempt it, rather do a little Google searching and report what you find.)
\\ \textbf{Solution:} \\
It seems impossible to obtain MLEs of the parameters in the model.
\part Now suppose $n\ll T_i$ for $i\in\{A,B\}$.  While the model you proposed above is still valid, what other model with fewer parameters is now plausible for these data?
\\ \textbf{Solution:} \\
Since $n \ll T_i$, the sampling procedure can be consider as with replacement. 
\[X_{ij} \sim B(n, \theta_{ij}),\]
where $B(\cdot)$ is binomial distribution and $\theta_{ij}$ is the unknown proportion of transcript $j$ among all transcripts in cell $i$.
\part What is $H_0$ for this model?
\\ \textbf{Solution:} \\
Still, 
\[H_0: \theta_{Aj} = \theta_{Bj}\]
\part Another convenient test for hypotheses about population parameters is the likelihood ratio test (LRT).
It is useful when the null hypothesis imposes constraints on parameters in the more general model of the alternative hypothesis.
Read about it in the Stat 342 notes or anywhere else (there are lots of resources; it is a common test) and implement it for this problem, first generically for any data $X_{Aj}, X_{Bj}$, and then specifically for $n=100, X_{Aj}=3$, and $X_{Bj} = 10$.
\\ \textbf{Solution:} \\
\[H_0: \theta_{Aj} = \theta_{Bj}\]
\[H_1: \theta_{Aj} \ne \theta_{Bj}\]
\begin{equation*}
\begin{split}
\lambda & = \frac{\max_{\theta_{Aj} = \theta_{Bj}}L(\Theta; X)}{\max_{\Theta} L(\Theta;X)} \\
   &= \frac{\max_{\theta} [\theta^{X_{Aj} + X_{Bj}} (1-\theta)^{2n-X_{Aj}-X_{Bj}} ]}
{\max_{\theta_{Aj},\theta_{Bj}} [\theta_{Aj}^{X_{Aj}}\theta_{Bj}^{X_{Bj}} (1-\theta_{Aj})^{n-X_{Aj}} (1-\theta_{Bj})^{n-X_{Bj}}]} \\
   &= \frac{ (\frac{X_{Aj}+X_{Bj}}{2n})^{X_{Aj} + X_{Bj}} (1-\frac{X_{Aj}+X_{Bj}}{2n})^{2n-X_{Aj}-X_{Bj}} }
{ (\frac{X_{Aj}}{n} )^{X_{Aj}} (\frac{X_{Bj}}{n})^{X_{Bj}} (1-\frac{X_{Aj}}{n})^{n-X_{Aj}} (1-\frac{X_{Bj}}{n})^{n-X_{Bj}}  } \\
\end{split}
\end{equation*}

<<>>==
a1 = 13/200;
h0=a1^(13)*(1-a1)^(200-13);
a2 = 3/100;
a3 = 10/100;
h1 = a2^3*(1-a2)^97*a3^10*(1-a3)*90;
log.lambda = log(h0/h1);
pchisq ( -2*log.lambda, df=1, lower.tail=F );

@

\part Argue that the Law of Rare Events (or the Poisson limit theorem, according to Wikipedia) applies in this case, and reformulate your model as a Poisson model.
\\ \textbf{Solution:}\\
Since $n$ is large and $\theta$ is small, Poisson distribution is a reasonable approximation of the binomial distribution. 
\[X_{ij} \sim P(\theta_{ij})\]

\part Spend about a paragraph explaining what conclusions you can draw when $H_0$ is rejected.
If you can reject $H_0$ because the $p$-value (see Stat 342 notes if you need assistance) is very tiny, what can you really conclude?
How much closer are you to identifying the cause of this type of cancer?
Think both biologically and statistically.
\\ \textbf{Solution:} \\
If we can reject $H_0$, we can say it is very unlikely to observe the data we have observed under $H_0$ and alternatively $H_1$ is true. 
However, we are still far from knowing the cause of this type of cancer. The statistical concern is that we have not yet measured the variability between cells.  It is plausible that we could identify just as many DE genes by comparing two type A cells because of natural variation between cells.  Thus, the DE genes supported by the test may not only be a consequence of the cancer, but may have nothing to do with cancer at all. Even if we assume the model is true and we identify the real DE genes, i.e., we know which transcripts were differentially transcribed but it still might be the outcome of the cancer instead of the reason. 
\end{parts}

\end{questions}

\end{document}
